### PR TITLE
CATROID-1071 Refactor GoToRandomPositionAction to Kotlin

### PR DESCRIPTION
**Pull request by mpozvek**
https://jira.catrob.at/browse/CATROID-1071

Refactoring of the class GoToRandomPositionAction and the associated tests to Kotlin.

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [ ] Include the name of the Jira ticket in the PR’s title
- [ ] Include a summary of the changes plus the relevant context
- [ ] Choose the proper base branch (*develop*)
- [ ] Confirm that the changes follow the project’s coding guidelines
- [ ] Verify that the changes generate no compiler or linter warnings
- [ ] Perform a self-review of the changes
- [ ] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Confirm that new and existing unit tests pass locally
- [ ] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [ ] Stick to the project’s gitflow workflow
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/GoToRandomPositionActionTest.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@RunWith(JUnit4::class)
+class GoToRandomPositionActionTest {
+    @Rule @JvmField
+    val exception : ExpectedException? = ExpectedException.none()
+
+    private var sprite: Sprite? = null
+    private var dummySprite: Sprite? = null
```
Comment (blindguardian50) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/GoToRandomPositionActionTest.kt:43]: Please make the members lateinit and not nullable so you can avoid !! operator

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/GoToRandomPositionActionTest.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            sprite!!.look.xInUserInterfaceDimensionUnit
+        )
+        Assert.assertEquals(
+            action!!.getRandomYPosition(),
+            sprite!!.look.yInUserInterfaceDimensionUnit
+        )
+    }
+
+    @Test
+    fun testNullActor() {
```
Comment (blindguardian50) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/GoToRandomPositionActionTest.kt:77]: I think this is not a very nice test as it expects the code to throw a null pointer exception which not gets caught and we surely not want that to happen. Instead you could give your sprite a position and check if the position has not changed if the passed sprite is null.

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/GoToRandomPositionActionTest.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            sprite, dummySprite, BrickValues.GO_TO_RANDOM_POSITION
+        ) as? GoToRandomPositionAction
+    }
+
+    @Test
+    fun testGoToOtherSpriteAction() {
+        sprite!!.look.xInUserInterfaceDimensionUnit = 0f
+        sprite!!.look.yInUserInterfaceDimensionUnit = 0f
+
+        Assert.assertEquals(0f, sprite!!.look.xInUserInterfaceDimensionUnit)
```
Comment (blindguardian50) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/GoToRandomPositionActionTest.kt:61]: Again you can save codelines here if you import the member instead of the whole Assert class

**Review by blindguardian50 — CHANGES_REQUESTED** [review]
Good work! Please consider my suggested changes

File: catroid/src/main/java/org/catrobat/catroid/content/actions/GoToRandomPositionAction.kt [hunk 4]
```diff
@@ -20,40 +20,23 @@
...
+package org.catrobat.catroid.content.actions
+
+import com.badlogic.gdx.scenes.scene2d.actions.TemporalAction
+import org.catrobat.catroid.common.ScreenValues
+import org.catrobat.catroid.content.Sprite
+
+class GoToRandomPositionAction : TemporalAction() {
+    var sprite: Sprite? = null
+    var randomXPosition: Float = Math.random().toFloat() * (ScreenValues.SCREEN_WIDTH + 1) -
+        (ScreenValues.SCREEN_WIDTH / 2)
```
Comment (blindguardian50) [inline catroid/src/main/java/org/catrobat/catroid/content/actions/GoToRandomPositionAction.kt:32]: Please consider my suggestion to import the members SCREEN_WIDTH and SCREEN_HEIGHT directly using ALT + ENTER and then import members from ...

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/GoToRandomPositionActionTest.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        assertEquals(0f, sprite.look.yInUserInterfaceDimensionUnit)
+
+        action.act(1f)
+
+        assertEquals(action.randomXPosition, sprite.look.xInUserInterfaceDimensionUnit)
+        assertEquals(action.randomYPosition, sprite.look.yInUserInterfaceDimensionUnit)
+    }
+
+    @Test
+    fun testCheckPositionOnNullSpriteAfterAction() {
```
Comment (blindguardian50) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/GoToRandomPositionActionTest.kt:66]: Instead of using a nullSprite which is not null, you could acutally use null for the sprite parameter when creating the GoToRandomPositionAction. Then you could execute your action in your test and if no exception occurs the behaviour is OK.

**Review by blindguardian50 — CHANGES_REQUESTED** [review]
Good Work! After adapting my suggesting changes I will approve! Please also have a look at the static code analysis, there are also a few style issues.

File: catroid/src/main/java/org/catrobat/catroid/content/actions/GoToRandomPositionAction.kt [hunk 6]
```diff
@@ -20,40 +20,22 @@
...
+import org.catrobat.catroid.content.Sprite
+
+class GoToRandomPositionAction : TemporalAction() {
+    var sprite: Sprite? = null
+    var randomXPosition: Float = Math.random().toFloat() * SCREEN_WIDTH + 1 - SCREEN_WIDTH / 2
+    var randomYPosition: Float = Math.random().toFloat() * SCREEN_HEIGHT + 1 - SCREEN_HEIGHT / 2
+
+    override fun update(percent: Float) {
+        sprite?.apply {
+            look.xInUserInterfaceDimensionUnit = randomXPosition
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/content/actions/GoToRandomPositionAction.kt:37]: I could be wrong, but I think this can't work that way. 
randomXPosition and randomYPosition won't be updated.
Have you tried using the respective brick within a loop?

Comment (gPathpp) [reply catroid/src/main/java/org/catrobat/catroid/content/actions/GoToRandomPositionAction.kt:37]: If you inject the "Math.random" method, you could test corner cases ;)
Also, it does not do the same as the prev. method, you forgot one parenthesis :) But I think it doesn't make a big difference

**Review by gPathpp — CHANGES_REQUESTED** [review]

File: catroid/src/main/java/org/catrobat/catroid/content/actions/GoToRandomPositionAction.kt (hunk 6 above)
Comment (mpozvek) [reply catroid/src/main/java/org/catrobat/catroid/content/actions/GoToRandomPositionAction.kt:37]: You are absolutely right in your assumption. randomXPosition and randomYPosition won't be updated. I will fix it and inform you afterwards.

**Comment by reichli** [comment]
closed due to inactivity
